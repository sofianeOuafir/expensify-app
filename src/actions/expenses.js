import uuid from "uuid";
import database from "../firebase/firebase";

const addExpense = expense => ({
  type: "ADD_EXPENSE",
  expense
});

const removeExpense = ({ id }) => ({
  type: "REMOVE_EXPENSE",
  id
});

const editExpense = ({ id, updates }) => ({
  type: "EDIT_EXPENSE",
  id,
  updates
});

const startEditExpense = ({ id, updates }) =>  {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/expenses/${id}`).update(updates).then(() => {
      dispatch(editExpense({id, updates}));
    });
  }
}

const startAddExpense = (expenseData = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    const {
      description = "",
      note = "",
      amount = 0,
      createdAt = 0
    } = expenseData;
    const expense = { description, note, amount, createdAt };
    return database
      .ref(`users/${uid}/expenses`)
      .push(expense)
      .then(ref => {
        dispatch(
          addExpense({
            id: ref.key,
            ...expense
          })
        );
      })
      .catch(e => {
        console.log(e);
      });
  };
};

const setExpenses = expenses => ({
  type: "SET_EXPENSES",
  expenses
});

const startSetExpenses = () => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database
      .ref(`users/${uid}/expenses`)
      .once("value")
      .then(snapshot => {
        let expenses = [];
        snapshot.forEach(expense => {
          expenses.push({
            id: expense.key,
            ...expense.val()
          });
        });
        dispatch(setExpenses(expenses));
      });
  };
};

const startRemoveExpense = ({ id }) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/expenses/${id}`).remove().then(() => {
      dispatch(removeExpense({ id }));
    });
  }
}

export {
  addExpense,
  removeExpense,
  editExpense,
  startEditExpense,
  startAddExpense,
  startSetExpenses,
  setExpenses,
  startRemoveExpense
};
