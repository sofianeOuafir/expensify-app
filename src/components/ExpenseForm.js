import React from 'react';
import { SingleDatePicker } from 'react-dates';
import 'react-dates/lib/css/_datepicker.css';
import moment from 'moment';

class ExpenseForm extends React.Component {
  constructor (props) {
    super(props);
    this.state = {
      description: props.expense && props.expense.description || '',
      amount: props.expense && parseFloat(props.expense.amount / 100) || '',
      note: props.expense && props.expense.note || '',
      createdAt: props.expense && moment(props.expense.createdAt) || moment(),
      calendarFocused: false,
      error: ''
    };
  }

  onDescriptionChange = (e) => {
    const description = e.target.value;
    this.setState(() => ({ description }));
  };

  onAmountChange = (e) => {
    const amount = e.target.value;
    if(!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState(() => ({ amount }));
    }
  };

  onNoteChange = (e) => {
    const note = e.target.value;
    this.setState(() => ({ note }))
  };

  onDateChange = (createdAt) => { 
    if(createdAt){
      this.setState(() => ({ createdAt }));
    }
  };

  onFocusChange = ({focused}) => { this.setState(() => ({ calendarFocused: focused })) };

  onSubmit = (e) => {
    e.preventDefault();
    if(this.state.description && this.state.amount){
      this.props.onSubmit({
        description: this.state.description,
        amount: parseFloat(this.state.amount) * 100,
        createdAt: this.state.createdAt.valueOf(),
        note: this.state.note
      });
      this.setState(() => ({ error: '' }));
    } else {
      this.setState(() => ({ error: 'Please provide description and amount.' }));
    }
  };

  render () {
    return (
    <div>
      {this.state.error && <p>{this.state.error}</p>}
      <form onSubmit={this.onSubmit}>
        <input 
          type="text"
          placeholder="Description"
          autoFocus
          value={this.state.description} 
          onChange={this.onDescriptionChange} 
        />
        <input 
          type="text" 
          placeholder="Amount"
          value={this.state.amount}
          onChange={this.onAmountChange}
        />
        <SingleDatePicker 
          date={this.state.createdAt}
          onDateChange={ this.onDateChange }
          focused={ this.state.calendarFocused }
          onFocusChange={ this.onFocusChange }
          numberOfMonths={1}
          isOutsideRange={(day) => false}
        />
        <textarea
          placeholder="Add a note for your expense (optional)"
          value={this.state.note}
          onChange={this.onNoteChange}
        >
        </textarea>
        <button>
          Save Expense
        </button>
      </form>
    </div>
    )
  }
}

export default ExpenseForm;
