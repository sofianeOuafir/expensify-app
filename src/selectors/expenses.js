import moment from 'moment';

const getVisibleExpenses = (expenses, {text, sortBy, startDate, endDate}) => {
  return expenses.filter((expense) => {
    const matchStartDate = startDate ? startDate.isSameOrBefore(expense.createdAt, 'day') : true;
    const matchEndDate = endDate ? endDate.isSameOrAfter(expense.createdAt, 'day') : true;
    const matchText = expense.description.toLowerCase().includes(text.toLowerCase());
    return matchText && matchStartDate && matchEndDate;
  }).sort((a, b) => {
    if (sortBy === 'date'){
      return a.createdAt < b.createdAt ? 1 : -1;
    } else if (sortBy === 'amount'){
      return a.amount < b.amount ? 1 : -1;
    }
  });
};

export { getVisibleExpenses };