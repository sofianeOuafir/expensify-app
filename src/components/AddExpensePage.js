import React from 'react';
import ExpenseForm from './ExpenseForm';
import { addExpense } from '../actions/expenses';
import { connect } from 'react-redux';

class AddExpensePage extends React.Component {
  render () {
    return (
      <div>
        <h1>Add Expense</h1>
        <ExpenseForm
          onSubmit={(expense) => {
            this.props.dispatch(addExpense(expense));
            this.props.history.push('/');
          }}
        />
      </div>
    );
  }
}

export default connect()(AddExpensePage);
