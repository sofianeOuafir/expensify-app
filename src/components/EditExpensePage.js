import React from "react";
import ExpenseForm from "./ExpenseForm";
import { connect } from "react-redux";
import { startEditExpense, startRemoveExpense } from "../actions/expenses";
import Modal from 'react-modal';

export class EditExpensePage extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      modalIsOpen: false
    };
  }

  openModal = () => {
    this.setState(() => ({modalIsOpen: true}));
  }

  closeModal = () => {
    this.setState(() => ({modalIsOpen: false}));;
  }

  editExpense = expense => {
    this.props.editExpense(this.props.expense.id, expense).then(() => {
      this.props.history.push("/dashboard");
    });
  };

  removeExpense = () => {
    this.props.removeExpense(this.props.expense.id).then(() => {
      this.props.history.push("/dashboard");
    });
  };
  render() {
    return (
      <div>
        <div className="page-header">
          <div className="content-container">
            <h1 className="page-header__title">Edit Expense</h1>
          </div>
        </div>
        <div className="content-container">
          <ExpenseForm
            expense={this.props.expense}
            onSubmit={this.editExpense}
          />
          <button className="button button--remove" onClick={this.openModal}>Remove Expense</button>
          <Modal
            className="modal"
            isOpen={this.state.modalIsOpen}
            onRequestClose={this.closeModal}
            contentLabel="Remove Expense"
          >
            <h1 className="modal__title">Remove Expense</h1>
            <p className="modal__content">Are you sure you want to remove this expense?</p>
            <div className="modal__button-group">
              <button className="button button--remove" onClick={this.removeExpense}>Remove</button>
              <button className="button" onClick={this.closeModal}>Cancel</button>
            </div>
          </Modal>
        </div>
      </div>
    );
  }
}

const mapStateToProps = (state, props) => {
  return {
    expense: state.expenses.find(
      expense => expense.id === props.match.params.id
    )
  };
};

const mapDispatchToProps = dispatch => {
  return {
    editExpense: (id, expense) => {
      return dispatch(
        startEditExpense({
          id,
          updates: expense
        })
      );
    },
    removeExpense: id => {
      return dispatch(
        startRemoveExpense({
          id
        })
      );
    }
  };
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(EditExpensePage);
