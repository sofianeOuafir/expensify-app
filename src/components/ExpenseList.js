import React from "react";
import { connect } from "react-redux";
import ExpenseListItem from "./ExpenseListItem";
import { getVisibleExpenses } from "../selectors/expenses";

export const ExpenseList = props => (
  <div className="content-container">
    <div className="list-header">
      <div className="show-for-mobile">Expenses</div>
      <div className="show-for-desktop">Expense</div>
      <div className="show-for-desktop">Amount</div>
    </div>
    <div className="list-item__body">
      {props.expenses.length === 0 ? (
        <div className="list-item list-item--message">
          <span>No expenses</span>
        </div>
      ) : (
        props.expenses.map((expense, index) => (
          <ExpenseListItem key={index} {...expense} />
        ))
      )}
    </div>
  </div>
);

const mapStateToProps = ({ expenses, filters }) => {
  return {
    expenses: getVisibleExpenses(expenses, filters)
  };
};

export default connect(mapStateToProps)(ExpenseList);
