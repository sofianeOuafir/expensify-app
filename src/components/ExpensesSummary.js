import React from "react";
import { connect } from "react-redux";
import selectExpensesTotal from '../selectors/expenses-total';
import numeral from 'numeral';
import { getVisibleExpenses } from '../selectors/expenses'; 

export const ExpensesSummary = ({ expenseCount, total }) => (
  <div>
    <h1>Viewing {expenseCount} {expenseCount && expenseCount > 1 ? 'expenses' : 'expense' } totalling {numeral(total / 100).format('$0,0.00')} </h1>
  </div>
);

const mapStateToProps = ({ expenses, filters }) => {
  const visibleExpenses = getVisibleExpenses(expenses, filters);
  return {
    expenseCount: visibleExpenses.length,
    total: selectExpensesTotal(visibleExpenses)
  }
};

export default connect(mapStateToProps)(ExpensesSummary);
