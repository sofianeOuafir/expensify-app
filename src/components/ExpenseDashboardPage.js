import React from 'react';
import ExpenseList from './ExpenseList';
import ExpenseListFilters from './ExpenseListFilters';
import ExpensesSummary from './ExpensesSummary';
import { connect } from 'react-redux';

const ExpenseDashboardPage = () => (
  <div>
    <ExpensesSummary />
    <ExpenseListFilters />
    <ExpenseList />
  </div>
);

export default ExpenseDashboardPage;
