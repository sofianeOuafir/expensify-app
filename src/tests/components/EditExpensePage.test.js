import React from "react";
import { shallow } from "enzyme";
import { EditExpensePage } from "../../components/EditExpensePage";
import expenses from '../fixtures/expenses';

let wrapper, editExpense, removeExpense, history;

beforeEach(() => {
  editExpense = jest.fn();
  removeExpense = jest.fn();
  history = { push: jest.fn() };

  wrapper = shallow(
    <EditExpensePage
      expense={expenses[0]}
      editExpense={editExpense}
      removeExpense={removeExpense}
      history={history}
    />
  );
});

test("should render correctly", () => {
  expect(wrapper).toMatchSnapshot();
});

test('should handle edit expense', () => {
  const expense = expenses[0];
  wrapper.find('ExpenseForm').simulate('submit', expense);
  expect(editExpense).toHaveBeenCalledWith(expense.id, expense);
  expect(history.push).toHaveBeenCalledWith('/');
});

test('should handle remove expense', () => {
  const id = expenses[0].id;
  wrapper.find('button').simulate('click');
  expect(removeExpense).toHaveBeenCalledWith(id);
  expect(history.push).toHaveBeenCalledWith('/');
});