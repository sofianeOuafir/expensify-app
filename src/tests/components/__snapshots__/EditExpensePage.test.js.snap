// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
<div>
  <div
    className="page-header"
  >
    <div
      className="content-container"
    >
      <h1
        className="page-header__title"
      >
        Edit Expense
      </h1>
    </div>
  </div>
  <div
    className="content-container"
  >
    <ExpenseForm
      expense={
        Object {
          "amount": 195,
          "createdAt": 0,
          "description": "Gum",
          "id": "1",
          "note": "",
        }
      }
      onSubmit={[Function]}
    />
    <button
      className="button button--remove"
      onClick={[Function]}
    >
      Remove Expense
    </button>
    <Modal
      ariaHideApp={true}
      bodyOpenClassName="ReactModal__Body--open"
      className="modal"
      closeTimeoutMS={0}
      contentLabel="Remove Expense"
      isOpen={false}
      onRequestClose={[Function]}
      parentSelector={[Function]}
      portalClassName="ReactModalPortal"
      shouldCloseOnOverlayClick={true}
    >
      <h1
        className="modal__title"
      >
        Remove Expense
      </h1>
      <p
        className="modal__content"
      >
        Are you sure you want to remove this expense?
      </p>
      <div
        className="modal__button-group"
      >
        <button
          className="button button--remove"
          onClick={[Function]}
        >
          Remove
        </button>
        <button
          className="button"
          onClick={[Function]}
        >
          Cancel
        </button>
      </div>
    </Modal>
  </div>
</div>
`;
